infrastructure/src/main/java/org/corfudb/infrastructure/CustomServerRouter.java

```diff
@@ -0,0 +1,186 @@
+    public void sendResponse(ChannelHandlerContext ctx, CorfuMsg inMsg, CorfuMsg outMsg) {
+        log.info("Ready to send response {}", outMsg.getMsgType());
+        outMsg.copyBaseFields(inMsg);
+        try {
+            serverAdapter.send(CorfuMessageConverter.toProtoBuf(outMsg));
+            log.trace("Sent response: {}", outMsg);
+        } catch (IllegalArgumentException e) {
+            log.warn("Illegal response type. Ignoring message.", e);
+        }
+    }
+
+    @Override
+    public Optional<Layout> getCurrentLayout() {
+        return Optional.empty();
+    }
+
+    @Override
+    public List<AbstractServer> getServers() {
+        return servers;
+    }
+
+    @Override
+    public void setServerContext(ServerContext serverContext) {
+
+    }
+
+    /**
+     * Receive messages from the 'custom' serverAdapter implementation. This message will be forwarded
+     * for processing.
+     *
+     * @param protoMessage
+     */
+    public void receive(CorfuMessage protoMessage) {
+        CorfuMsg corfuMsg;
+        try {
+            log.info("Received message {}", protoMessage.getType().name());
+            // Transform protoBuf into CorfuMessage
+            corfuMsg = CorfuMessageConverter.fromProtoBuf(protoMessage);
+        } catch (CorfuMessageProtoBufException e) {
+            log.error("Exception while trying to convert {} from protoBuf", protoMessage.getType(), e);
+            return;
+        }
+
+        AbstractServer handler = handlerMap.get(corfuMsg.getMsgType());
+        if (handler == null) {
+            // The message was unregistered, we are dropping it.
+            log.warn("Received unregistered message {}, dropping", corfuMsg);
+        } else {
+            if (validateEpoch(corfuMsg)) {
+                // Route the message to the handler.
+                if (log.isTraceEnabled()) {
+                    log.trace("Message routed to {}: {}", handler.getClass().getSimpleName(), corfuMsg);
+                }
+
+                try {
+                    handler.handleMessage(corfuMsg, null, this);
+                } catch (Throwable t) {
+                    log.error("channelRead: Handling {} failed due to {}:{}",
+                            corfuMsg != null ? corfuMsg.getMsgType() : "UNKNOWN",
+                            t.getClass().getSimpleName(),
+                            t.getMessage(),
+                            t);
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate the epoch of a CorfuMsg, and send a WRONG_EPOCH response if
+     * the server is in the wrong epoch. Ignored if the message type is reset (which
+     * is valid in any epoch).
+     *
+     * @param msg The incoming message to validate.
+     * @return True, if the epoch is correct, but false otherwise.
+     */
+    private boolean validateEpoch(CorfuMsg msg) {
+        long serverEpoch = getServerEpoch();
+        if (!msg.getMsgType().ignoreEpoch && msg.getEpoch() != serverEpoch) {
```

Comment:
minor - it is sometimes easier to understand conditions without negations.  Perhaps we can change this to 
msg.getMsgType().ignoreEpoch || epoch = serverEpoch:
  //return straightaway